File: 0235-lowest-common-ancestor-of-a-binary-search-tree/0235-lowest-common-ancestor-of-a-binary-search-tree.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode(int x) : val(x), left(NULL), right(NULL) {}
 * };
 */

class Solution {
public:
    TreeNode* lowestCommonAncestor(TreeNode* root, TreeNode* p, TreeNode* q) {
        if(!root) return root;
        if(root->val == p->val || q->val == root->val){
            return root;
        }
        TreeNode* lside = lowestCommonAncestor(root->left , p , q);
        TreeNode* rside = lowestCommonAncestor(root->right , p , q);
        if(lside && rside) return root;
        if(!lside) return rside;
        return lside;
    }
};

File: 0235-lowest-common-ancestor-of-a-binary-search-tree/0235-lowest-common-ancestor-of-a-binary-search-tree.py
# Definition for a binary tree node.
# class TreeNode(object):
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None

class Solution(object):
    def lowestCommonAncestor(self, root, p, q):
        if not root : return None
        if root.val == p.val or root.val== q.val:
            return root
        lh = self.lowestCommonAncestor(root.left , p , q)
        rh = self.lowestCommonAncestor(root.right , p , q)

        if lh and rh : return root
        if not lh: return rh
        return lh